src/qibo/models/encodings.py

```diff
@@ -395,27 +385,33 @@ def _generate_rbs_pairs(nqubits: int, architecture: str, **kwargs):
     """
 
     if architecture == "diagonal":
-        pairs_rbs = np.arange(nqubits)
+        pairs_rbs = np.arange(nqubits - 1, -1, -1)
         pairs_rbs = [[pair] for pair in zip(pairs_rbs[:-1], pairs_rbs[1:])]
 
     if architecture == "tree":
-        pairs_rbs = [[(0, int(nqubits / 2))]]
-        indexes = list(pairs_rbs[0][0])
-        for depth in range(2, int(math.log2(nqubits)) + 1):
-            pairs_rbs_per_depth = [
-                [(index, index + int(nqubits / 2**depth)) for index in indexes]
-            ]
-            pairs_rbs += pairs_rbs_per_depth
-            indexes = list(np.array(pairs_rbs_per_depth).flatten())
-
-    pairs_rbs = [
-        [(nqubits - 1 - a, nqubits - 1 - b) for a, b in row] for row in pairs_rbs
-    ]
+        depth = int(np.ceil(np.log2(nqubits)))
+        registers = [list(np.arange(nqubits - 1, -1, -1))]
+        pairs_rbs = []
+        for _ in range(depth):
+            new_registers, per_depth = [], []
+            for register in registers:
+                limit = (
+                    int(len(register) / 2)
+                    if len(register) % 2 == 0
+                    else int(len(register) / 2) + 1
+                )
+                new_registers.append(register[:limit])
+                new_registers.append(register[limit:])
+                if len(register[:limit]) + len(register[limit:]) >= 2:
```

Comment:
isn't this the same as
```suggestion
                if len(register) >= 2:
```